### ISSUE TITLE
Current status of support for building with GCC 11

### ISSUE THREAD

**Issue by danielnachun** [ispc/ispc#2287]
We have recently been working on building `ispc` on Linux in Homebrew (https://github.com/Homebrew/homebrew-core/pull/97642) and I am seeking clarification on the implications of building `ispc` with GCC 11.  Looking through some previous issues, it seemed that users were encountering build failures when using GCC, and even some versions of LLVM.  However when I tried simply overwriting `CMAKE_C_COMPILER` and `CMAKE_CXX_COMPILER` with the appropriate GCC executables, everything builds fine with no errors.  This is true even if I enable `ISPC_INCLUDE_EXAMPLES`, `ISPC_INCLUDE_TESTS` and `ISPC_INCLUDE_UTILS`.  We also have a small test we run after building that successfully compiles a `.ispc` file.

Given these results, are there any potential downsides to building `ispc` with GCC 11 if it compiles successfully?  I see there are few patches you still ship for LLVM 13, but those seem to be for `libLLVM`, not `clang`, and only appear to be needed for targeting the Xe GPU architecture.  If there's anything else we should be checking for please let us know.

**Comment by dbabokin** [comment]
It's recommended to build ISPC with the same compiler as LLVM libs. For earlier versions of GCC / LLVM there was a known problem with ABI incompatibility, which caused problem, but it was fixed. Note that we use C++ interface for LLVM libs, so it's more error prone with respect to C++ ABI bugs and typical C ABI.

For that reason we do the following: do self-build of LLVM, so the resulting binaries and libs are built with LLVM/clang itself and use this clang to build ISPC. That's why we don't test GCC - we assume clang is always available, as we depend on LLVM libs.

Also, as you mentioned, we apply a set of patches to LLVM. For LLVM 13 right now we have two patches, which are GPU specific. Most of our patches are back-porting of the patches in LLVM trunk for problems that we discovered too late to fix in already released LLVM. For GPU some of the patches are workarounds for bugs / problems that we not yet solved in trunk.

We are trying to cover all known problems which require LLVM patches with lit tests (run with `make check-all`), if they are all passing, then most likely you are good.

Bottom line - you can build ISPC with GCC11, but it's highly recommended to build LLVM libs and ISPC with the same compiler. We also recommend applying all the patches we have for specific LLVM version (and build LLVM as part of ISPC build), but for ISPC 1.17 with LLVM 13, you don't need any patches, if you don't build GPU.

**Comment by dbabokin** [comment]
Also, please note that official ISPC 1.17 release was build with LLVM 12. Even though building with LLVM 13 was tested and worked fine for CPU (but not for GPU).

**Comment by danielnachun** [comment]
Thank for the detailed and helpful response.  We are indeed using the same compiler to build ISPC and the LLVM libs - in both cases GCC 11.  We may eventually move to building LLVM using a profile guided optimization that would build  LLVM with `clang`, but it would still use `libstdc++` from GCC, and `ispc` would still be built with the same version of `libstdc++` as well.  So `ispc` will always be built with same C++ API/ABI as the LLVM libs.  

We had a lengthy discussion with LLVM upstream about the importance of matching the C++ API/ABI we use to **build** LLVM with the C++ API/ABI we use to build packages that **link** to the LLVM libs, so we're already quite strict about following this rule for other packages.  Some people find it confusing that we don't use `libc++` to build the LLVM libs on Linux, but doing so would require us build all other packages that link the LLVM libs with `libc++` which get messy very quickly.

My questions are answered here so feel free to close this unless you have further thoughts/questions/suggestions.  Thank you once again for the quick and very helpful response!

**Comment by dbabokin** [comment]
C++ lib is not an issue for us (as long as it's the same, otherwise it's most likely just blow up). And `stdlibc++` seems to be reasonable choice on Linux.

I'm care more about C++ ABI - LLVM currently using C++14 and going to move to C++17. And compilers might disagree in the way some aspects of new standards affect ABI. But if something goes wrong, it will most likely cause linking issue, not runtime problem. That's what we saw in the past.
